### DISCUSSION TITLE
Calculating part of ERI (ao2mo)

### DISCUSSION THREAD

**Discussion by AlexisRalli** [pyscf/pyscf#2489] [Q&A] [answered]
I am trying to obtain a subset of the ERI terms in the MO basis. Specifically the `eris.papa` and `eris.ppaa` matrices. These can be obtained through the CASSCF module; however, it seems to return everything with S1 symmetry. Is it possible to calculate these matrices with S8 and/or S4 symmetry and return them with these symmetries (such that the internal calc and output has a smaller memory requirement)?

Here is a small example of what I am currently doing:
```
from pyscf import gto, scf, mcscf

mol = gto.M(    
    atom = [('N', (0, 0, 0)),
            ('H', (-0.4417, 0.2906, 0.8711)),
            ('H', (0.7256, 0.6896, -0.1907)),
            ('H', (0.4875, -0.8701, 0.2089))],
    basis='sto-3g',
    # basis='ccpvdz',
    unit='A',
)

hf_obj = scf.RHF(mol)
hf_obj.verbose=1
hf_obj.max_memory = 10_000

S = hf_obj.get_ovlp()

dm_init = hf_obj.init_guess_by_atom()
f_ao = hf_obj.get_fock(dm=dm_init)
_, C = hf_obj.eig(f_ao, S)

casscf_obj  = mcscf.CASSCF(hf_obj, hf_obj.mol.nao, hf_obj.mol.nelec, ncore=0)
eris = casscf_obj.ao2mo(C)

print(eris.papa.shape, eris.ppaa.shape)

## how can eri.papa and eri.ppaa be calculated and output
## using known symmetries of the integrals
```

Thanks for any help!

I have been trying to modify code in`mc_ao2mo`:
```
from pyscf.mcscf.mc_ao2mo import _ao2mo
from pyscf import ao2mo, lib
import numpy as np

def modified_trans_e1_incore(eri_ao, C, aosym='s4', mosym='s1'):
    # modified: https://github.com/pyscf/pyscf/blob/master/pyscf/mcscf/mc_ao2mo.py#L39

    nao = C.shape[0]

    eri1 = ao2mo.incore.half_e1(eri_ao, (C,C), compact=False)
    eri1 = eri1.reshape(nao,nao,-1)

    klppshape = (0, nao, 0, nao)
    klpashape = (0, nao, 0, nao)
    papa = np.empty((nao,nao,nao,nao))
    for i in range(nao):
        _ao2mo.nr_e2(eri1[i,:], C, klppshape,
                      aosym=aosym, mosym=mosym, out=papa[i])
    ppaa = lib.transpose(papa.reshape(nao*nao,-1)).reshape(nao,nao,nao,nao)
    
    return papa, ppaa
    
eri_ao_S8 = hf_obj.mol.intor('int2e', aosym='s8')
ppaa, papa = modified_trans_e1_incore(eri_ao_S8, C)

print(np.allclose(eris.papa, papa), np.allclose(eris.ppaa, ppaa))
```
However, this seems to require `mosym='s1'`

```
## The following has too much info and has calculated unnecessary MO integrals
ao2mo_S4 = ao2mo.kernel(eri_ao_S8, C)
ao2mo_S8 = ao2mo.restore(8, ao2mo_S4, C.shape[0])
## instead, the goal is to have: ppaa_S4, papa_S4 and/or  ppaa_S8, papa_S8
```

**Comment by sunqm** [comment, marked as answer]
To reduce the complexity in mcscf code, ppaa and papa integrals are stored without considering symmetry. Typically, in a system with moderate number of orbitals, such as ~1000 MOs and ~10 active space, these two tensors do not occupy too much memory.

To compress these tensor, additional code has to be implemented within the mc_ao2mo function, https://github.com/pyscf/pyscf/blob/0d83af70a67d8a550886d54a6990dc93f114419b/pyscf/mcscf/mc_ao2mo.py#L194 .

There are no functions available in PySCF to automatically handle the symmetry in papa .

If you use the `ao2mo.general` function to compute ppaa, the four-fold symmetry can be utilized, when `mosym='s4'` is specified. The output tensor should contain only Nmo*(Nmo+1)/2 * Nact*(Nact+1)/2 elements.
